File: leetcode/0733-flood-fill/solution.py
class Solution:
    def floodFill(
        self, image: list[list[int]], sr: int, sc: int, color: int
    ) -> list[list[int]]:
        m = len(image)
        n = len(image[0])
        start_value = image[sr][sc]

        if start_value == color:
            return image

        def dfs(r, c):
            if image[r][c] != start_value:
                return
            
            image[r][c] = color

            for dr, dc in [(1,0),(-1,0),(0,1),(0,-1)]:
                nr, nc = r + dr, c + dc
                if 0 <= nr < m and 0 <= nc < n:
                    dfs(nr, nc)
        
        dfs(sr, sc)
        return image        

File: leetcode/0733-flood-fill/solution.rs
use std::collections::VecDeque;

impl Solution {
    pub fn flood_fill(mut image: Vec<Vec<i32>>, sr: i32, sc: i32, color: i32) -> Vec<Vec<i32>> {
        let h = image.len() as i32;
        let w = image[0].len() as i32;

        let starting_color = image[sr as usize][sc as usize];
        if starting_color == color {
            return image;
        }

        let mut queue = VecDeque::new();

        image[sr as usize][sc as usize] = color;
        queue.push_back((sc, sr));

        while let Some((x, y)) = queue.pop_front() {
            for (dx, dy) in [(1, 0), (-1, 0), (0, 1), (0, -1)] {
                let nx = x + dx;
                let ny = y + dy;

                if 0 <= nx
                    && nx < w
                    && 0 <= ny
                    && ny < h
                    && image[ny as usize][nx as usize] == starting_color
                {
                    image[ny as usize][nx as usize] = color;
                    queue.push_back((nx, ny));
                }
            }
        }

        image
    }
}
